File: Dotnet/Dotnet.Mqtt/utils/TreeNode.cs
using System.Collections;

namespace Dotnet.Mqtt.Utils;

public class TreeNode : IEnumerable<TreeNode>
{
    private readonly Dictionary<string, TreeNode> children = new();

    public readonly string Topic;
    public List<IMqttHandler> Handlers = new();
    public TreeNode? Parent { get; set; }

    public TreeNode(string topic)
    {
        this.Topic = topic;
    }

    public TreeNode GetChild(string topic)
    {
        return children[topic];
    }

    public List<TreeNode> GetChildren()
    {
        return children.Values.ToList();
    }

    public IEnumerator<TreeNode> GetEnumerator()
    {
        return children.Values.GetEnumerator();
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
        return GetEnumerator();
    }

    public int Count
    {
        get { return children.Count; }
    }

    public void AddHandler(IMqttHandler handler)
    {
        Handlers.Add(handler);
    }

    public void RemoveHandler(IMqttHandler handler)
    {
        Handlers.Remove(handler);
    }

    public List<IMqttHandler> GetHandlers()
    {
        return Handlers;
    }

    public TreeNode GetRoot()
    {
        var root = this;
        var isRoot = false;

        while (!isRoot)
        {
            if (root.Parent != null)
            {
                root = root.Parent;
            }
            else
            {
                isRoot = true;
            }
        }

        return root;
    }

    //Return the tree node according to the topic. There is always 0 or 1 path.
    public TreeNode? GetTreeNode(string topic)
    {
        var root = GetRoot();

        //split topic
        var topics = topic.Split('/');

        if (topics.Length == 0)
        {
            return null;
        }

        //Move through the tree according to the topic and return the treeNode.
        var current = root;

        foreach (var t in topics)
        {
            if (current.children.ContainsKey(t))
            {
                current = current.GetChild(t);
            }
            else
            {
                return null;
            }
        }

        return current;
    }

    public void AddTreeNode(string topic, IMqttHandler? handler)
    {
        var topics = topic.Split('/');

        if (topics.Length == 0)
        {
            return;
        }

        var current = this;

        foreach (var t in topics)
        {
            if (current.children.ContainsKey(t))
            {
                current = current.GetChild(t);
            }
            else
            {
                var newNode = new TreeNode(t)
                {
                    Parent = current
                };
                
                current.children.Add(t, newNode);
                current = newNode;
            }
        }

        if (handler != null)
        {
            current.AddHandler(handler);
        }
    }

    //Return all handlers that are subscribed to the topic
    public List<IMqttHandler>? GetMqttHandlers(string topic)
    {
        TreeNode root = GetRoot();

        var topics = topic.Split('/');

        if (topics.Length == 0)
        {
            return null;
        }

        List<IMqttHandler> handlers = new();
        List<TreeNode> current = new() { root };

        foreach (var t in topics)
        {
            List<TreeNode> next = new();

            foreach (var c in current)
            {
                if (c.children.ContainsKey(t))
                {
                    next.Add(c.GetChild(t));
                }

                //skip 1 level
                if (c.children.ContainsKey("+"))
                {
                    next.Add(c.GetChild("+"));
                }

                //endpoint (you will never have ../#/... as topic)
                if (c.children.ContainsKey("#"))
                {
                    handlers.AddRange(c.GetChild("#").GetHandlers());
                }
            }

            current = next;
        }

        current.ForEach(c => handlers.AddRange(c.GetHandlers()));

        return handlers;
    }

    public List<string> GetAllSubscriptions(TreeNode? node)
    {
        List<string> subscriptions = new();

        var treeNode = node ?? GetRoot();

        foreach (var child in treeNode.children)
        {
            //only add subscriptions that have handlers
            if(child.Value.Handlers.Count > 0)
            {
                subscriptions.Add(child.Key);
            }

            if (child.Value.Count > 0)
            {
                var childSubscriptions = child.Value.GetAllSubscriptions(child.Value);

                foreach (var childSubscription in childSubscriptions)
                {
                    subscriptions.Add(child.Key + "/" + childSubscription);
                }
            }
        }

        return subscriptions;
    }
}

File: Nodejs/web-app/src/mqtt/utils/treeNode.ts
import IMqttHandler from "../mqttHandlers/IMqttHandler";

export default class TreeNode{
    private children: Map<string, TreeNode> = new Map();
    public topic: string;
    public handlers: IMqttHandler[] = [];
    public parent?: TreeNode;

    constructor(topic: string) {
        this.topic = topic;
    }

    public getChild(topic: string): TreeNode | undefined {
        return this.children.get(topic);
    }

    public getChildren(): TreeNode[] {
        return Array.from(this.children.values());
    }

    public count(): number {
        return this.children.size;
    }

    public addHandler(handler: IMqttHandler): void {
        this.handlers.push(handler);
    }

    public removeHandler(handler: IMqttHandler): void {
        const index = this.handlers.indexOf(handler);
        if (index !== -1) {
            this.handlers.splice(index, 1);
        }
    }

    public getHandlers(): IMqttHandler[] {
        return this.handlers;
    }

    public getRoot(): TreeNode {
        let root: TreeNode = this;
        let isRoot = false;

        while (!isRoot) {
            if(root.parent) {
                root = root.parent;
            } else {
                isRoot = true;
            }
        }
        
        return root;
    }

    public getTreeNode(topic: string): TreeNode | undefined {
        const root = this.getRoot();
        const topics = topic.split('/');

        if (topics.length === 0) {
            return undefined;
        }

        let current: TreeNode|undefined = root;

        for (const t of topics) {
            if(current?.children.has(t)) {
                current = current.getChild(t);
            } else {
                return undefined;
            }
        }

        return current;
    }

    public addTreeNode(topic: string, handler?: IMqttHandler): void {
        const topics = topic.split('/');

        if (topics.length === 0) {
            return;
        }

        let current: TreeNode|undefined = this;

        for (const t of topics) {
            if(current?.children.has(t)) {
                current = current.getChild(t);
            } else {
                const newNode = new TreeNode(t);
                newNode.parent = current;
                current?.children.set(t, newNode);
                current = newNode;
            }
        }

        if (handler) {
            current?.addHandler(handler);
        }
    }

    public getMqttHandlers(topic: string): IMqttHandler[] | null {
        const root = this.getRoot();
        const topics = topic.split('/');

        if (topics.length === 0) {
            return null;
        }

        let handlers: IMqttHandler[] = [];
        let current: TreeNode[] = [root];

        for (const t of topics) {
            const next: TreeNode[] = [];

            for (const c of current) {
                if (c.children.has(t)) {
                    const tmp = c.getChild(t);
                    if (tmp) next.push(tmp);
                }

                if (c.children.has("+")) {
                    const tmp = c.getChild("+");
                    if (tmp) next.push(tmp);
                }

                if (c.children.has("#")) {
                    const tmp = c.getChild("#");
                    if (tmp) handlers.push(...tmp.getHandlers());
                }
            }

            current = next;
        }

        for (const c of current) {
            handlers.push(...c.getHandlers());
        }

        return handlers;
    }

    public getAllSubscriptions(node?: TreeNode): string[] {
        const subscriptions: string[] = [];
        const treeNode = node ?? this.getRoot();

        for (const [key, child] of treeNode.children) {
            // Only add subscriptions that have handlers
            if (child.handlers.length > 0) {
                subscriptions.push(key);
            }

            if (child.count() > 0) {
                const childSubscriptions = child.getAllSubscriptions(child);

                for (const childSubscription of childSubscriptions) {
                    subscriptions.push(`${key}/${childSubscription}`);
                }
            }
        }

        return subscriptions;
    }
}
